Newpassives

GabeHasWon opened the pull request:
﻿### Link Issues
Resolves: #94 

### Description of Work
Adds Damage Reduction passive ("Iron Will") on the old passive system before #75. Will need to be updated before merging.

### Comments

CollinHerber commented:
The names of the files shouldn't change, also going to convert to draft so I'm not email blasted. Mark ready for review once ready

GabeHasWon commented:
You mean the passive icons? I renamed them to be consistent with class names (i.e. DamageReductionPassive instead of DamagereductionPassive) since Windows and Git are both capitalization agnostic and that can cause issues.

CollinHerber commented:
> You mean the passive icons? I renamed them to be consistent with class names (i.e. DamageReductionPassive instead of DamagereductionPassive) since Windows and Git are both capitalization agnostic and that can cause issues.

They need to be consistent with the InternalIdentifier ,  as this is how the PassiveTreeEditor finds the images and generates the json

GabeHasWon commented:
The only things not in the format LikeThisPassive are the images; the identifiers, classes and any other references are properly capitalized. This is just more consistent (and it looks nicer!)
I know on Linux and Mac an inconsistency like that will actually be picked up by the OS ("DamagereductionPassive" would be unable to load since it's not "DamageReductionPassive").

CollinHerber commented:
> The only things not in the format LikeThisPassive are the images; the identifiers, classes and any other references are properly capitalized. This is just more consistent (and it looks nicer!) I know on Linux and Mac an inconsistency like that will actually be picked up by the OS ("DamagereductionPassive" would be unable to load since it's not "DamageReductionPassive").

The image name, and casing must match the InternalIdentifier

You didn't just change the casing, but rather the others as well
![image](https://github.com/Path-of-Terraria/PathOfTerraria/assets/24656209/93aed658-7578-4ffb-a80d-ef4ebe655302)

When you make the DamageReductionPassive class just make sure the image name matches the InternalIdentifier.

GabeHasWon commented:
Oh, I must've merged those wrong. I didn't intend to change those, my bad. The only intended changes were capitalization fixes.

GabeHasWon commented:
Fixed the bad names, and there's a few other things to note:
`Passive.Draw` now uses `InternalIdentifier` to grab the associated texture instead of GetType().Name, the latter of which simply never worked under the current system.
I've also added `Passive.OnLoad` and `TreePlayer.GetCumulativeLevel`, such that I can subscribe to a `PathOfTerrariaPlayerEvents` event easily & get the player's current real level for a given passive. `OnLoad` is called on the dummy instance in `Passive.LoadPassives`.

Otherwise, #94 is done - it's hard to tell, but it works - and #93 is done. The only thing that may be off about #93 is that it is affected by defense, i.e. an NPC with 20 defense will take noticeably less damage than another with less defense.

CollinHerber reviewed (approved):
This looks good to me! Nicely done
